pandas/core/arrays/boolean.py

```diff
@@ -740,6 +741,160 @@ def boolean_arithmetic_method(self, other):
+
+    result = left | right
+    mask[left & ~left_mask] = False
+    if right_mask is not None:
+        mask[right & ~right_mask] = False
+    elif right is True:
+        mask[:] = False
+
+    return result, mask
+
+
+def kleene_xor(
+    left: Union[bool, np.ndarray],
+    right: Union[bool, np.ndarray],
+    left_mask: Optional[np.ndarray],
+    right_mask: Optional[np.ndarray],
+):
+    """
+    Boolean ``xor`` using Kleene logic.
+
+    This is the same as ``or``, with the following adjustments
+
+    * True, True -> False
+    * True, NA   -> NA
+
+    Parameters
+    ----------
+    left, right : ndarray, NA, or bool
+        The values of the array.
+    left_mask, right_mask : ndarray, optional
+        The masks. When
+
+    Returns
+    -------
+    result, mask: ndarray[bool]
+        The result of the logical xor, and the new mask.
+    """
+    if left_mask is None:
+        return kleene_xor(right, left, right_mask, left_mask)
+
+    raise_for_nan(right, method="xor")
+    # Re-use or, and update with adjustments.
+    result, mask = kleene_or(left, right, left_mask, right_mask)
+
+    # # TODO(pd.NA): change to pd.NA
+    # if lib.is_scalar(right) and right is libmissing.NA:
+    #     # True | NA == True
+    #     # True ^ NA == NA
+    #     mask[result] = True
+
+    result[left & right] = False
+    mask[right & left_mask] = True
+    if right_mask is not None:
+        mask[left & right_mask] = True
+
+    result[mask] = False
+    return result, mask
+
+
+def kleene_and(
+    left: Union[bool, np.ndarray],
+    right: Union[bool, np.ndarray],
+    left_mask: Optional[np.ndarray],
+    right_mask: Optional[np.ndarray],
+):
+    """
+    Boolean ``and`` using Kleene logic.
+
+    Values are ``NA`` for ``NA & NA`` or ``True & NA``.
+
+    Parameters
+    ----------
+    left, right : ndarray, NA, or bool
+        The values of the array.
+    left_mask, right_mask : ndarray, optional
+        The masks. When
+
+    Returns
+    -------
+    result, mask: ndarray[bool]
+        The result of the logical xor, and the new mask.
+    """
+    # To reduce the number of cases, we ensure that `left` & `left_mask`
+    # always come from an array, not a scalar. This is safe, since because
+    # A | B == B | A
+    if left_mask is None:
+        return kleene_and(right, left, right_mask, left_mask)
+
+    assert isinstance(left, np.ndarray)
+    raise_for_nan(right, method="and")
+
+    if right_mask is None:
+        # Scalar `right`
+        # TODO(pd.NA): handle NA here.
+        result = left & right
+
+        mask = left_mask.copy()
+        if right is False:
+            # unmask everything
+            mask[:] = False
+    else:
+        result = left & right
+        # unmask where either left or right is False
+        left_false = ~(left | left_mask)
+        right_false = ~(right | right_mask)
+        mask = (left_mask & ~right_false) | (right_mask & ~left_false)
+
+    result[mask] = False
```

Comment:
I don't think we need to do this (we can't rely on that in general anyway, and then this just gives a performance degradation)